services/src/main/scala/cromwell/services/metadata/impl/WriteMetadataActor.scala

```diff
@@ -1,29 +1,96 @@
+    super.preStart()
+  }
+
+  startWith(WaitingToWrite, NoEvents)
+
+  when(WaitingToWrite) {
+    case Event(PutMetadataAction(events), curData) =>
+      curData.addEvents(events) match {
+        case data@HasEvents(e) if e.size > batchRate => goto(WritingToDb) using data
+        case e => stay using e
+      }
+    case Event(ScheduledFlushToDb, curData) =>
+      log.debug("Initiating periodic metadata flush to DB")
+      goto(WritingToDb) using curData
+  }
+
+  when(WritingToDb) {
+    case Event(ScheduledFlushToDb, curData) => stay using curData
+    case Event(PutMetadataAction(events), curData) => stay using curData.addEvents(events)
+    case Event(FlushBatchToDb, curData) =>
+      curData match {
+        case NoEvents =>
+          log.debug("Attempted metadata flush to DB but had nothing to write")
+          goto(WaitingToWrite) using curData
+        case HasEvents(e) =>
+          log.debug("Flushing {} metadata events to the DB", e.size)
+          addMetadataEvents(e) onComplete {
+            case Success(_) => self ! DbWriteComplete
+            case Failure(regerts) =>
+              log.error("Failed to properly flush metadata to database", regerts)
+              self ! DbWriteComplete
+          }
+
+          stay using NoEvents
+      }
+    case Event(DbWriteComplete, curData) =>
+      log.debug("Flush of metadata events complete")
+      goto(WaitingToWrite) using curData
+  }
+
+  onTransition {
+    case WaitingToWrite -> WritingToDb => self ! FlushBatchToDb
+  }
 }
 
-class WriteMetadataActor extends Actor with ActorLogging with MetadataDatabaseAccess with SingletonServicesStore {
+object WriteMetadataActor {
+  def props(batchRate: Int, flushRate: FiniteDuration) = Props(new WriteMetadataActor(batchRate, flushRate)).withDispatcher(ServiceDispatcher)
 
-  implicit val ec = context.dispatcher
+  sealed trait WriteMetadataActorMessage
+  case object DbWriteComplete extends WriteMetadataActorMessage
+  case object FlushBatchToDb extends WriteMetadataActorMessage
+  case object ScheduledFlushToDb extends WriteMetadataActorMessage
+
+  sealed trait WriteMetadataActorState
+  case object WaitingToWrite extends WriteMetadataActorState
+  case object WritingToDb extends WriteMetadataActorState
 
-  def receive = {
-    case action@PutMetadataAction(events) =>
-      val sndr = sender()
-      addMetadataEvents(events) onComplete {
-        case Success(_) => sndr ! MetadataPutAcknowledgement(action)
-        case Failure(t) =>
-          val msg = MetadataPutFailed(action, t)
-          log.error(t, "Sending {} failure message {}", sndr, msg)
-          sndr ! msg
+  sealed trait WriteMetadataActorData {
+    def addEvents(newEvents: Iterable[MetadataEvent]): WriteMetadataActorData = {
+      if (newEvents.isEmpty) this
+      else {
+        val newEventsVector = newEvents.toVector
+        val combinedEvents: Vector[MetadataEvent] = this match {
+          case NoEvents => newEventsVector
```

Comment:
FWIW I like the theory here, but only if `HasEvents` enforces a non-empty value. Otherwise we aren't really using the type safety.
Similarly, if `newEvents` was types as non-empty, we shouldn't need the extra `if/else` checks here